gradio/app.py

```diff
@@ -0,0 +1,314 @@
+
+
+@app.get("/api", response_class=HTMLResponse)  # Needed for Spaces
+@app.get("/api/", response_class=HTMLResponse)
+def api_docs(request: Request):
+    inputs = [type(inp) for inp in app.interface.input_components]
+    outputs = [type(out) for out in app.interface.output_components]
+    input_types_doc, input_types = get_types(inputs, "input")
+    output_types_doc, output_types = get_types(outputs, "output")
+    input_names = [type(inp).__name__ for inp in app.interface.input_components]
+    output_names = [type(out).__name__ for out in app.interface.output_components]
+    if app.interface.examples is not None:
+        sample_inputs = app.interface.examples[0]
+    else:
+        sample_inputs = [inp.generate_sample() for inp in app.interface.input_components]
+    docs = {
+        "inputs": input_names,
+        "outputs": output_names,
+        "len_inputs": len(inputs),
+        "len_outputs": len(outputs),
+        "inputs_lower": [name.lower() for name in input_names],
+        "outputs_lower": [name.lower() for name in output_names],
+        "input_types": input_types,
+        "output_types": output_types,
+        "input_types_doc": input_types_doc,
+        "output_types_doc": output_types_doc,
+        "sample_inputs": sample_inputs,
+        "auth": app.interface.auth,
+        "local_login_url": urllib.parse.urljoin(
+            app.interface.local_url, "login"),
+        "local_api_url": urllib.parse.urljoin(
+            app.interface.local_url, "api/predict")
+    }
+    return templates.TemplateResponse(
+        "api_docs.html", 
+        {"request": request, **docs}
+    )
+
+
+@app.post("/api/predict/", dependencies=[Depends(login_check)])
+async def predict(
+    request: Request, 
+    username: str = Depends(get_current_user)
+):
+    body = await request.json()
+    flag_index = None
+    
+    if body.get("example_id") != None:
+        example_id = body["example_id"]
+        if app.interface.cache_examples:
+            prediction = load_from_cache(app.interface, example_id)
+            durations = None
+        else:
+            prediction, durations = process_example(app.interface, example_id)
+    else:
+        raw_input = body["data"]
+        if app.interface.show_error:
+            try:
+                prediction, durations = app.interface.process(raw_input)
+            except BaseException as error:
+                traceback.print_exc()
+                return JSONResponse(content={"error": str(error)}, 
+                                    status_code=500)
+        else:
+            prediction, durations = app.interface.process(raw_input)
+        if app.interface.allow_flagging == "auto":
+            flag_index = app.interface.flagging_callback.flag(
+                app.interface, raw_input, prediction,
+                flag_option="" if app.interface.flagging_options else None, 
+                username=username)
+    output = {
```

Comment:
here as well, using response model validations would be beneficial

Reply:
Good idea